src/reader.rs

```diff
@@ -88,31 +43,29 @@ where
     C: Read + Write,
 {
     async fn fill_buf(&mut self) -> Result<&[u8], ErrorKind> {
-        // We need to consume the loaded bytes before we read mode.
-        if self.buffer.is_empty() {
-            // embedded-tls has its own internal buffer, let's prefer that if we can
-            #[cfg(feature = "embedded-tls")]
-            if let HttpConnection::Tls(ref mut tls) = self.stream {
+        // The call to buffered.bypass() will only return Ok(...) if the buffer is empty.
+        // This ensures that we completely drain the possibly pre-filled buffer before we try
+        // to use the embedded-tls buffer directly.
+        // The matches/if let dance is to fix lifetime of the borrowed inner connection.
+        #[cfg(feature = "embedded-tls")]
+        if matches!(self.buffered.bypass(), Ok(HttpConnection::Tls(_))) {
+            if let HttpConnection::Tls(ref mut tls) = self.buffered.bypass().unwrap() {
```

Comment:
`bypass` is a neat idea!

So do I understand correctly that `if let Ok(HttpConnection::Tls(ref mut tls)) = self.buffered.bypass() {` without the outer `matches!` would not work˙, even if you wrap the whole block in a new scope?

Reply:
Year, I have tried multiple versions, which all fail to compile. The problem is that `self.buffered` is still borrowed in the else branch, and I cannot figure out how to avoid that. Any good proposals?

Reply:
Is it possible to change `.bypass()` to return `Result<&mut T, &mut Self>`?

Reply:
If you think that would make it work, then yes:)

Reply:
I guess I can try locally first :)

Reply:
Well, no, it can't because we still need to handle the `HttpConnection::Plain` branch :(